include/common/array.h

```diff
@@ -30,4 +30,15 @@ wl_array_len(struct wl_array *array)
 	return array->size / sizeof(const char *);
 }
 
+/**
+ * Iterates in reverse over an array.
+ * @pos: pointer that each array element will be assigned to
+ * @array: wl_array to iterate over
+ */
+#define wl_array_for_each_reverse(pos, array)                                          \
+	for (pos = !(array)->data ? NULL                                               \
+		: (void *)((const char *)(array)->data + (array)->size - sizeof(pos)); \
+		pos && (const char *)pos >= (const char *)(array)->data;               \
+		pos--)
```

Comment:
A few more eyes on this would be great

Reply:
Macro looks good. I've played with it, and it seems to work well.

It's unfortunate that we have to deal with the `!data` case and also that we have to involve `sizeof(pos)`, but I think you've come up with the simplest way.

We should enclose the final pos in brackets (`(pos)--`). Partly because it's good practice to do that for parameters in macros but also to be consistent with the `wl_array_for_each()` macro.

Reply:
changed